File: app/src/main/java/id/ac/uasbungaa/ui/activity/barang/Detail200020016BarangActivity.java
package id.ac.uasbungaa.ui.activity.barang;

import android.app.Activity;
import android.content.ContentValues;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;

import id.ac.uasbungaa.R;
import id.ac.uasbungaa.data.local.BarangHelper;
import id.ac.uasbungaa.data.local.DatabaseContract;
import id.ac.uasbungaa.data.model.Barang;
import id.ac.uasbungaa.databinding.ActivityDetail200020016BarangBinding;
import id.ac.uasbungaa.helper.Keys;
import id.ac.uasbungaa.helper.ToastHelper;

import java.io.ByteArrayOutputStream;

import kotlin.jvm.internal.Intrinsics;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class Detail200020016BarangActivity extends AppCompatActivity {
    private ActivityDetail200020016BarangBinding binding;
    private Barang barang;
    private BarangHelper barangHelper;
    private String path = "";
    private View overlay;
    @NotNull
    public static final String EXTRA_BARANG = "EXTRA_BARANG";

    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityDetail200020016BarangBinding.inflate(this.getLayoutInflater());
        this.setContentView(binding.getRoot());

        barang = this.getIntent().getParcelableExtra(EXTRA_BARANG);
        ActionBar supportActionBar = this.getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.setTitle(barang.getNama());
            supportActionBar.setDisplayHomeAsUpEnabled(true);
        }

        overlay = this.findViewById(R.id.progbar_layout);
        barangHelper = new BarangHelper(this);

        Comparable imgId;
        if (barang.getGambar().contains("flower")) {
            imgId = this.getResources().getIdentifier(
                    barang.getGambar(),
                    "drawable",
                    this.getPackageName()
            );
        } else {
            imgId = Uri.parse(barang.getGambar());
        }
        Glide.with(this)
                .load(imgId)
                .into(binding.civThumbnail);
        binding.etKode.setText(barang.getKode());
        binding.etNama.setText(barang.getNama());
        binding.etHarga.setText(String.valueOf(barang.getHarga()));

        binding.civThumbnail.setOnClickListener((OnClickListener) (new OnClickListener() {
            public final void onClick(View it) {
                checkStoragePermission();
            }
        }));

        binding.btnSubmit.setOnClickListener((OnClickListener) (new OnClickListener() {
            public final void onClick(View it) {
                overlay.setVisibility(View.VISIBLE);
                handleSubmit();
            }
        }));
    }

    private final void checkStoragePermission() {
        if (this.checkSelfPermission("android.permission.WRITE_EXTERNAL_STORAGE") != PackageManager.PERMISSION_GRANTED) {
            this.requestPermissions(new String[]{"android.permission.WRITE_EXTERNAL_STORAGE"}, Keys.WRITE_EXTERNAL_CODE);
        } else {
            this.checkCameraPermission();
        }
    }

    private final void checkCameraPermission() {
        if (this.checkSelfPermission("android.permission.CAMERA") != PackageManager.PERMISSION_GRANTED) {
            this.requestPermissions(new String[]{"android.permission.CAMERA"}, Keys.CAMERA_PERMISSION_CODE);
        } else {
            Intent intent = new Intent("android.media.action.IMAGE_CAPTURE");
            this.startActivityForResult(intent, Keys.CAMERA_REQUEST);
        }
    }

    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == Keys.CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            Bitmap photo = (Bitmap) data.getExtras().get("data");
            photo.compress(
                    CompressFormat.JPEG,
                    100,
                    new ByteArrayOutputStream()
            );
            path = MediaStore.Images.Media.insertImage(
                    this.getContentResolver(),
                    photo,
                    barang.getNama() + "_Image",
                    null
            );
            binding.civThumbnail.setImageURI(Uri.parse(path));
        }
    }

    public void onRequestPermissionsResult(int requestCode, @NotNull String[] permissions, @NotNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == Keys.CAMERA_PERMISSION_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.INSTANCE.showToast("Camera Permission Granted", this);
                Intent intent = new Intent("android.media.action.IMAGE_CAPTURE");
                this.startActivityForResult(intent, Keys.CAMERA_REQUEST);
            } else {
                ToastHelper.INSTANCE.showToast("Camera Permission Denied", this);
            }
        }

        if (requestCode == Keys.WRITE_EXTERNAL_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.INSTANCE.showToast("Storage Permission Granted", this);
                this.checkCameraPermission();
            } else {
                ToastHelper.INSTANCE.showToast("Storage Permission Denied", this);
            }
        }
    }

    public boolean onOptionsItemSelected(@NotNull MenuItem item) {
        Intrinsics.checkNotNullParameter(item, "item");
        switch (item.getItemId()) {
            case R.id.btn_delete:
                this.handleDeleteBarang();
                break;
            case android.R.id.home:
                this.finish();
        }

        return super.onOptionsItemSelected(item);
    }

    public boolean onCreateOptionsMenu(@Nullable Menu menu) {
        this.getMenuInflater().inflate(R.menu.menu_user, menu);
        return super.onCreateOptionsMenu(menu);
    }

    private final void handleSubmit() {
        String kode = binding.etKode.getText().toString();
        String nama = binding.etNama.getText().toString();
        String harga = binding.etHarga.getText().toString();
        updateRecord(kode, nama, Integer.parseInt(harga));
        finish();
    }

    private final void updateRecord(String kode, String nama, int harga) {
        barangHelper.open();
        ContentValues values = new ContentValues();
        if (path != "") {
            values.put(DatabaseContract.BARANG_200020016.GAMBAR, path);
        }
        values.put(DatabaseContract.BARANG_200020016.KODE, kode);
        values.put(DatabaseContract.BARANG_200020016.NAMA, nama);
        values.put(DatabaseContract.BARANG_200020016.HARGA, harga);
        barangHelper.update(String.valueOf(barang.getId()), values);
        barangHelper.close();
    }

    private final void handleDeleteBarang() {
        barangHelper.open();
        barangHelper.delete(String.valueOf(barang.getId()));
        barangHelper.close();
        finish();
    }
}

File: app/src/main/java/id/ac/uasbunga/ui/activity/Detail200020016BarangActivity.kt
package id.ac.uasbunga.ui.activity

import android.Manifest
import android.app.Activity
import android.content.ContentValues
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.MediaStore
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.bumptech.glide.Glide
import id.ac.uasbunga.R
import id.ac.uasbunga.data.local.BarangHelper
import id.ac.uasbunga.data.local.DatabaseContract.BARANG_200020016.Companion.GAMBAR
import id.ac.uasbunga.data.local.DatabaseContract.BARANG_200020016.Companion.HARGA
import id.ac.uasbunga.data.local.DatabaseContract.BARANG_200020016.Companion.KODE
import id.ac.uasbunga.data.local.DatabaseContract.BARANG_200020016.Companion.NAMA
import id.ac.uasbunga.data.model.Barang
import id.ac.uasbunga.databinding.ActivityDetail200020016BarangBinding
import id.ac.uasbunga.helper.Keys
import id.ac.uasbunga.helper.ToastHelper
import java.io.ByteArrayOutputStream

class Detail200020016BarangActivity : AppCompatActivity() {
    companion object {
        const val EXTRA_BARANG = "EXTRA_BARANG"
    }

    private lateinit var binding: ActivityDetail200020016BarangBinding
    private lateinit var barang: Barang
    private lateinit var barangHelper: BarangHelper
    private var path = ""
    private lateinit var overlay: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDetail200020016BarangBinding.inflate(layoutInflater)
        setContentView(binding.root)

        barang = intent.getParcelableExtra(EXTRA_BARANG)!!
        supportActionBar?.title = barang.nama
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        overlay = findViewById(R.id.progbar_layout)
        barangHelper = BarangHelper.getInstance(this)

        val gambar = if (barang.gambar.contains("flower"))
            resources.getIdentifier(
                barang.gambar,
                "drawable",
                packageName
            )
        else
            Uri.parse(barang.gambar)
        Glide.with(this).load(gambar).into(binding.civThumbnail)
        binding.etKode.setText(barang.kode)
        binding.etNama.setText(barang.nama)
        binding.etHarga.setText(barang.harga.toString())

        binding.civThumbnail.setOnClickListener {
            checkStoragePermission()
        }

        binding.btnSubmit.setOnClickListener {
            overlay.visibility = View.VISIBLE
            handleSubmit()
        }
    }

    private fun checkStoragePermission() {
        if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                Keys.WRITE_EXTERNAL_CODE
            )
        } else {
            checkCameraPermission()
        }
    }

    private fun checkCameraPermission() {
        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.CAMERA), Keys.CAMERA_PERMISSION_CODE)
        } else {
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            startActivityForResult(intent, Keys.CAMERA_REQUEST)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == Keys.CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            val photo = data?.extras?.get("data") as Bitmap
            photo.compress(Bitmap.CompressFormat.JPEG, 100, ByteArrayOutputStream())
            path =
                MediaStore.Images.Media.insertImage(
                    this.contentResolver,
                    photo,
                    "${barang.nama}_Image",
                    null
                )
            binding.civThumbnail.setImageURI(Uri.parse(path))
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == Keys.CAMERA_PERMISSION_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.showToast("Camera Permission Granted", this)
                val intent = Intent(android.provider.MediaStore.ACTION_IMAGE_CAPTURE)
                startActivityForResult(intent, Keys.CAMERA_REQUEST)
            } else {
                ToastHelper.showToast("Camera Permission Denied", this)
            }
        }
        if (requestCode == Keys.WRITE_EXTERNAL_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ToastHelper.showToast("Storage Permission Granted", this)
                checkCameraPermission()
            } else {
                ToastHelper.showToast("Storage Permission Denied", this)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.btn_delete -> handleDeleteBarang()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_user, menu)
        return super.onCreateOptionsMenu(menu)
    }

    private fun handleSubmit() {
        val kode = binding.etKode.text.toString()
        val nama = binding.etNama.text.toString()
        val harga = binding.etHarga.text.toString()
        updateRecord(kode, nama, Integer.parseInt(harga))
        finish()
    }

    private fun updateRecord(kode: String, nama: String, harga: Int) {
        barangHelper.open()
        val values = ContentValues()
        if (path != "") values.put(GAMBAR, path)
        values.put(KODE, kode)
        values.put(NAMA, nama)
        values.put(HARGA, harga)
        barangHelper.update(barang.id.toString(), values)
        barangHelper.close()
    }

    private fun handleDeleteBarang() {
        barangHelper.open()
        barangHelper.delete(barang.id.toString())
        barangHelper.close()
        finish()
    }
}